### PR TITLE
Solved issue with applying shopping cart rule to a single bundle item

### LINKED ISSUE #28921: Apply shopping cart rule to a single bundle item
### Preconditions (*)
Magento ver. 2.3.5-p1

### Steps to reproduce (*)
1. Create a bundle product with two option each contain two product in it
2. Create one option as main as the title with two simple product (a&b) - (Dropdown)
3. Create the second option as a sub as the title with two simple product(c&d)- (Checkbox)
4. Create a shopping cart rule 
5. Select a "percent of product price discount" 
6. Add 10 % to the discount amount 
7. We should add the main option two product in the rule like( SKU is one of (a&b)
8. Apply the rule 
9. Go to the frontend
10. Select one SKU from the dropdown of the main simple product and checkbox one sub simple product 
11. Add to cart
12. Go to cart page
13. Check the discount is applied total of two product

### Expected result (*)
10% discount is applied ONLY on BUNDLE child product SKU001 (this behavior worked faultless on Magento 1.8.X)
The discount should be applied only for the main simple product 
### Actual result (*)
10% discount is applied on BUNDLE parent product, thus both of simple products are discounted...
bundle_issue

Additional information
more detailed steps and screens #28921 ([Comment1](https://github.com/magento/magento2/issues/28921#issuecomment-651136449)) ([comment 2](https://github.com/magento/magento2/issues/28921#issuecomment-651167895))

### PR DESCRIPTION
**Pull request by zhartaunik**
### Description (*)
There are two kinds of parent products: Bundle and Configurable (Grouped does not present in the quote)
For the Configurable product - we should calculate discount for the parent.
For the Bundle product - we should calculate children and ignore the Parent.

### Fixed Issues (if relevant)
1. Fixes magento/magento2#28921

### Manual testing scenarios (*)
1. Create a bundle product with two option each contain two product in it
2. Create one option as main as the title with two simple product (a&b) - (Dropdown)
3. Create the second option as a sub as the title with two simple product(c&d)- (Checkbox)
4. Create a shopping cart rule
5. Select a "percent of product price discount"
6. Add 10 % to the discount amount
7. We should add the main option two product in the rule like( SKU is one of (a&b)
8. Apply the rule
9. Go to the frontend
10. Select one SKU from the dropdown of the main simple product and checkbox one sub simple product
11. Add to cart
12. Go to cart page
13. Check the discount is applied total of two product

### Questions or comments
All assumptions are in the description section

### Contribution checklist (*)
 - [*] Pull request has a meaningful description of its purpose
 - [*] All commits are accompanied by meaningful commit messages
 - [*] All new or changed code is covered with unit/integration tests (if applicable)
 - [*] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by zhartaunik** [comment]
What is `Expected — Waiting for status to be reported` ? Who should add this status?

**Comment by lbajsarowicz** [comment]
@magento run all tests

@zhartaunik Please adjust the PR title to reflect the goal of the change :-)

**Comment by zhartaunik** [comment]
Hi @lbajsarowicz 
There are two issues in static tests:

Coupling 14>13 (Please advice the better way: to replace constants with string values or add @ignore) ?
Dependency issue. Depends on the previous point. To replace or not. Should I add to module.xml only or in composer.json too?

**Comment by lbajsarowicz** [comment]
@zhartaunik I'm not sure which solution would be the best one, however - for now, let's just add Ignore syntax to commit the issue, in the meantime I'll try to catch up with some Architect to make sure.

**Comment by zhartaunik** [comment]
@lbajsarowicz thanks for the suggestion.
Also not clear for me one more thing. My created MFTF test fails. It contains several rows sequence which creates CatalogRule. I took it from one of tests. There are at least 8 places where Magento uses the same sequence of steps in acceptance tests:
```
<actionGroup ref="SelectNotLoggedInCustomerGroupActionGroup" stepKey="selectNotLoggedInCustomerGroup"/>
<generateDate date="-1 day" format="m/d/Y" stepKey="yesterdayDate"/>
<fillField selector="{{AdminCartPriceRulesFormSection.fromDate}}" userInput="{$yesterdayDate}" stepKey="fillFromDate"/>
<click selector="{{AdminCartPriceRulesFormSection.actionsHeader}}" stepKey="clickToExpandActions"/>
```
But only in my case, it cannot find "fillFromDate" element. I suppose it happens because this element is out fo view. I tried to find any description regarding "codeception scrollTo" function. And this is not clear when to use it and where not to use it.

Do you have access to test results? Maybe you can help with some clarification?

**Comment by zhartaunik** [comment]
@lbajsarowicz 
I did the following:
 - Improved discount calculation algorithm. Now we split bundle products into two groups: with dynamic and fixed price. When the price is fixed - we ignore discount rules for child products as prices here are custom. And we apply a discount - to the parent product. On the contrary for dynamic price. Parent product doesn't have a price and consist of child product prices. In that case, we calculate discount only for children.
 - I decided to make 3 separate IF conditions. When we have one IF - it becomes really unreadable.
 - I've added dependencies and ignore coupling to pass static tests. In my opinion, there is no easy way to decrease coupling. However, I can suggest splitting logic for each separate product type and make sth like "validationPool"
 - I improved the MFTF test which repeats the same steps as inside described issue. **However, it still has a problem, which is described in my previous message.** The same sequence of steps works well in other tests, but in my, they fail. Any ideas why it could occur.
 - The issue with API-functional test should be also solved
 - I improved the unit test, added by me before. But please draw your attention at my new test and the approach I chose.

File: app/code/Magento/SalesRule/Model/Quote/Discount.php [hunk 1]
```diff
@@ -137,8 +141,26 @@ public function collect(
                 }
                 continue;
             }
-            // to determine the child item discount, we calculate the parent
-            if ($item->getParentItem()) {
+
+            /**
```
Comment (zhartaunik) [inline app/code/Magento/SalesRule/Model/Quote/Discount.php:157]: I decided to add 3 separate IFs
As for me, it looks more clear

File: app/code/Magento/SalesRule/Test/Mftf/Test/CartPriceRuleForBundleProductTest.xml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+        <!-- Create the rule -->
+        <amOnPage url="{{AdminCartPriceRulesPage.url}}" stepKey="amOnCartPriceList"/>
+        <waitForPageLoad stepKey="waitForRulesPage"/>
+        <click selector="{{AdminCartPriceRulesSection.addNewRuleButton}}" stepKey="clickAddNewRule"/>
+        <fillField selector="{{AdminCartPriceRulesFormSection.ruleName}}" userInput="{{SimpleSalesRule.name}}" stepKey="fillRuleName"/>
+        <selectOption selector="{{AdminCartPriceRulesFormSection.websites}}" userInput="Main Website" stepKey="selectWebsites"/>
+        <actionGroup ref="SelectNotLoggedInCustomerGroupActionGroup" stepKey="selectNotLoggedInCustomerGroup"/>
+        <generateDate date="-1 day" format="m/d/Y" stepKey="yesterdayDate"/>
+        <fillField selector="{{AdminCartPriceRulesFormSection.fromDate}}" userInput="{$yesterdayDate}" stepKey="fillFromDate"/>
```
Comment (zhartaunik) [inline app/code/Magento/SalesRule/Test/Mftf/Test/CartPriceRuleForBundleProductTest.xml]: At the 1st time, it failed here. Don't know why. Same code presents in other 8-9 tests.

**Comment by zhartaunik** [comment]
@magento run all tests

**Comment by zhartaunik** [comment]
@magento run all tests

**Comment by bubasuma** [comment]
@zhartaunik could you please confirm that this issue is still reproducible in 2.4-develop?
We fixed this issue recently here 71bb134ba3cd0b45da7d7e7c857ae2cdd412b05b
I tested the scenario in the ticket, and it works as expected in 2.4-develop. Am I missing something?

File: app/code/Magento/SalesRule/Test/Mftf/Test/CartPriceRuleForBundleProductTest.xml [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            <!-- Make Attribute 'sku' accessible for Promo Rule Conditions -->
+            <actionGroup ref="NavigateToEditProductAttributeActionGroup" stepKey="editSkuAttribute">
+                <argument name="ProductAttribute" value="sku" />
+            </actionGroup>
+            <actionGroup ref="ChangeUseForPromoRuleConditionsProductAttributeActionGroup" stepKey="changeAttributePromoRule">
+                <argument name="option" value="1" />
+            </actionGroup>
+
+            <!-- Run reindex stock status -->
+            <magentoCLI command="cron:run --group=index" stepKey="runCronIndexer"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/SalesRule/Test/Mftf/Test/CartPriceRuleForBundleProductTest.xml]:
```suggestion
            <magentoCron groups="index" stepKey="runCronIndexer" />
```

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Just a minor change to MFTF -- in the meantime : )
`<magentoCron` has an additional mechanism to keep the threshold between runs.

**Comment by zhartaunik** [comment]
@magento run all tests

**Comment by zhartaunik** [comment]
> @zhartaunik could you please confirm that this issue is still reproducible in 2.4-develop?
> We fixed this issue recently here [71bb134](https://github.com/magento/magento2/commit/71bb134ba3cd0b45da7d7e7c857ae2cdd412b05b)
> I tested the scenario in the description, and it works as expected in 2.4-develop. Am I missing something?

In my opinion, discount for bundle product with a dynamic price for bundle product itself does not make sense.

There is an option "Dynamic price" for bundle products.

Dynamic price YES:
Bundle product always has zero price. The price field is blocked. Bundle product is just a container.
Child products take their original price.

Dynamic price NO:
Bundle product must have price > 0
Child products ignoring their regular price. You must set a surcharge for them. If you leave them empty it means that surcharge equals zero.

**Comment by zhartaunik** [comment]
> @zhartaunik could you please confirm that this issue is still reproducible in 2.4-develop?
> We fixed this issue recently here [71bb134](https://github.com/magento/magento2/commit/71bb134ba3cd0b45da7d7e7c857ae2cdd412b05b)
> I tested the scenario in the description, and it works as expected in 2.4-develop. Am I missing something?

Thanks for the reply, I dunno why, but I missed notification about your response.
I responded in the main issue.

File: app/code/Magento/SalesRule/Model/Quote/Discount.php (hunk 1 above)
Comment (lbajsarowicz) [reply app/code/Magento/SalesRule/Model/Quote/Discount.php:157]: You can actually extract that to separate private method. This way code would be more SOLID

Comment (zhartaunik) [reply app/code/Magento/SalesRule/Model/Quote/Discount.php:157]: @lbajsarowicz thanks for the suggestion. It looks like the issue was already solved and in PR mentioned by @bubasuma 
And I'm thinking about reverting all changes and leaving only MFTF test if community is interested in it.

Comment (lbajsarowicz) [reply app/code/Magento/SalesRule/Model/Quote/Discount.php:157]: I see! Leaving MFTF sounds reasonable. Thank you for your effort.

**Comment by zhartaunik** [comment]
Hi @lbajsarowicz 
cc @bubasuma 

This PR is supposed to be closed without merge. Here is PR with the only MFTF test: https://github.com/magento/magento2/pull/29348

**Comment by lbajsarowicz** [comment]
@zhartaunik OK. Closing the issue as per request.

**Comment by bubasuma** [comment]
@zhartaunik thanks so much. I think the MFTF coverage is perfect. Thanks again for your collaboration.

**Pull request closed without merging**